src/parse.c

```diff
@@ -768,11 +768,33 @@ handle_netdef_map(NetplanParser* npp, yaml_node_t* node, const char* key_prefix,
     return handle_generic_map(npp, node, key_prefix, npp->current.netdef, data, error);
 }
 
+/*
+ * Check if the passthrough key format is incorrect and remove it from the list.
+ * user_data is expected to contain a pointer to the GData list.
+ */
+static void
+validate_kf_group_key(GQuark key_id, gpointer value, gpointer user_data)
+{
+    GData** list = user_data;
+    const gchar* key = g_quark_to_string(key_id);
+    gchar** group_key = g_strsplit(key, ".", -1);
+    if (g_strv_length(group_key) < 2) {
+        g_warning("NetworkManager: passthrough key '%s' format is invalid, should be 'group.key'.", key);
+        g_datalist_id_remove_data(list, key_id);
+    }
+    g_strfreev(group_key);
+}
+
 static gboolean
-handle_netdef_datalist(NetplanParser* npp, yaml_node_t* node, const char* key_prefix, const void* data, GError** error)
+handle_passthrough_datalist(NetplanParser* npp, yaml_node_t* node, const char* key_prefix, const void* data, GError** error)
 {
     g_assert(npp->current.netdef);
-    return handle_generic_datalist(npp, node, key_prefix, npp->current.netdef, data, error);
+    gboolean ret = handle_generic_datalist(npp, node, key_prefix, npp->current.netdef, data, error);
```

Comment:
I didn't want to touch handle_generic_datalist() or write a more specific version of it to not duplicate code. Although it would be more efficient that way.

My solution validated the passthrough words that were loaded from the YAML and remove the ones that doesn't have at least 2 components separated by a dot. According to the glib documentation, it is safe to remove the key inside a foreach. https://docs.gtk.org/glib/func.datalist_foreach.html

Let me know what you think!

Reply:
Your approach wfm! It's good to warn the user when an entry is deleted/ignored.

But I think the function should be called `handle_netdef_passthrough_datalist` and we'd need a similar modification to `handle_access_point_datalist`. Also, we should think about what happens if the invalid entry was the only entry in the datalist. In this case, we should probably clear the datalist alltogether?

Consider extending the testcase like this:
```diff
--- a/tests/generator/test_passthrough.py
+++ b/tests/generator/test_passthrough.py
@@ -332,6 +332,14 @@ method=auto
 
     def test_passthrough_empty_keyfile_group(self):
         out = self.generate('''network:
+  wifis:
+    wlan0:
+      access-points:
+        "SSID":
+          networkmanager:
+            name: connection_name
+            passthrough:
+              itsmissingadot: abc
   nm-devices:
     device0:
       networkmanager:
```

Reply:
So, your new test case revealed another memory leak... I did something I was trying to avoid: I changed `reset_backend_settings()` to reset all the settings without checking what is the backend.

In your test for example, the key `networkmanager.name` would never be freed if the backend is not NetworkManager. So I just removed the `switch (){case}` block.